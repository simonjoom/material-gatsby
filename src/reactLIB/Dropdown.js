import React, { Component } from 'react';
import PropTypes from 'prop-types';
import idgen from './idgen';
import cx from 'classnames';

const classes = {
  'dropdown-content': true
};

class Dropdown extends Component {
  constructor(props) {
    super(props);
    this.idx = 'dropdown_' + idgen();
    this.renderTrigger = this.renderTrigger.bind(this);
  }

  componentDidMount() {
    const options = this.props.options || {}; 
    Zepto(this._trigger).dropdown(options); 
  }

  componentWillUnmount() {
    Zepto(this._trigger).off();
  }

  render() {
    const { children, className, ...props } = this.props;
    delete props.trigger;
    delete props.options;

    return (
      <span>
        {this.renderTrigger()}
        <ul {...props} className={cx(classes, className)} id={this.idx}>
          {children}
        </ul>
      </span>
    );
  }

  renderTrigger() {
    const { trigger } = this.props;

    return React.cloneElement(trigger, {
      ref: t => (this._trigger = `[data-target=${this.idx}]`),
      className: cx(trigger.props.className, 'dropdown-trigger'),
      'data-target': this.idx
    });
  }
}

Dropdown.propTypes = {
  /**
   * The node to trigger the dropdown
   */
  trigger: PropTypes.node.isRequired,
  children: PropTypes.node,
  className: PropTypes.string,

  /**
   * Options hash for the dropdown
   * <a target="_blank" href="http://materializecss.com/dropdown.html#options">http://materializecss.com/dropdown.html</a>
   */
  options: PropTypes.shape({
    inDuration: PropTypes.number,
    outDuration: PropTypes.number,
    constrainWidth: PropTypes.bool,
    hover: PropTypes.bool,
    gutter: PropTypes.number,
    coverTrigger: PropTypes.bool,
    alignment: PropTypes.oneOf(['left', 'right'])
  })
};

export default Dropdown;
