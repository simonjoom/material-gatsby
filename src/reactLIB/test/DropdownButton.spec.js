import React from 'react';
import { shallow } from 'enzyme';
import Dropdown from '../src/Dropdown';
import mocker from './helper/mocker';

describe('<Dropdown />', () => {
  let wrapper;
  const dropdownMock = jest.fn();
  const restore = mocker('dropdown', dropdownMock);

  afterAll(() => {
    restore();
  });

  test('renders', () => {
    wrapper = shallow(<Dropdown trigger={<button>Drop me!</button>} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('handles options prop', () => {
    const options = { constrainWidth: true };
    wrapper = shallow(
      <Dropdown trigger={<button>Drop me!</button>} options={options} />
    );
    expect(dropdownMock).toHaveBeenCalledWith(options);
  });

  test('deletes unwatend props', () => {
    const options = { hover: true };
    wrapper = shallow(<Dropdown trigger={<span>hi</span>} options={options} />);
    expect(wrapper.find('ul').props()).not.toMatchObject(options);
  });

  test('does not update ID reference with each render', () => {
    wrapper = shallow(<Dropdown trigger={<span>hi</span>} />);
    const { idx } = wrapper.instance();
    wrapper.render();
    expect(wrapper.instance().idx).toEqual(idx);
  });
});
