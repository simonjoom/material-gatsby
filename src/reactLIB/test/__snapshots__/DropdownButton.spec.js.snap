// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> renders 1`] = `
<span>
  <button
    className="dropdown-trigger"
    data-target="dropdown_0"
  >
    Drop me!
  </button>
  <ul
    className="dropdown-content"
    id="dropdown_0"
  />
</span>
`;
